### PR TITLE
Delete endpoint

### PR DESCRIPTION
**Pull request by hadrienk**
Linked PRs: 

- https://github.com/statisticsnorway/data-access-protobuf/pull/7
- https://github.com/statisticsnorway/data-access/pull/14
- https://github.com/statisticsnorway/dapla-cli/pull/9
- https://github.com/statisticsnorway/dapla-catalog/pull/23
- https://github.com/statisticsnorway/dapla-catalog/pull/24

### CODE REVIEW

File: src/main/java/no/ssb/dapla/datamaintenance/model/DeleteResponse.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package no.ssb.dapla.datamaintenance.model;
+
+import java.nio.file.Path;
+import java.time.Instant;
+import java.util.List;
+
+public record DeleteResponse(String datasetPath, List<DatasetVersion> deletedVersions) {
```
Comment (takvamborgen) [inline src/main/java/no/ssb/dapla/datamaintenance/model/DeleteResponse.java:7]: Should we add a dry-run flag to this, to show the caller if the result of the call indeed was a dry-run?

Comment (takvamborgen) [inline src/main/java/no/ssb/dapla/datamaintenance/model/DeleteResponse.java:7]: Are you certain about using preview features? Enabling of preview has caused problems with the pipelines previously.

File: src/main/java/no/ssb/dapla/datamaintenance/service/DataMaintenanceService.java [hunk 2]
```diff
@@ -143,17 +145,41 @@ public static DatasetListElement toDataset(CatalogClient.Identifier identifier)
...
+        // Make sure the dataset is not also a folder.
+        Single<Map<Identifier, DeleteLocationResponse>> tokens = catalogService.getDatasetVersions(datasetPath, Integer.MAX_VALUE)
+                .flatMap(identifier ->
+                        dataAccessService.getDeleteToken(identifier.path, identifier.timestamp, JWT)
+                                .map(r -> new AbstractMap.SimpleEntry<>(identifier, r)))
+                .collect(HashMap::new, (m, e) -> m.put(e.getKey(), e.getValue()));
+
+        if (!catalogService.isOnlyDataset(datasetPath, now).await()) {
+            tokens.cancel();
+            throw new HttpException("the path " + datasetPath + " is also a folder");
```
Comment (takvamborgen) [inline src/main/java/no/ssb/dapla/datamaintenance/service/DataMaintenanceService.java:164]: Do we not allow deletion if there exists a folder with the same name as the dataset? We then force the user to first delete the dataset(s) in the folder with the same name (and then recreate it somewhere else or after the deletion of the dataset with the same name as the folder), even if that is not desirable.

File: src/main/java/no/ssb/dapla/datamaintenance/storage/PathAndSize.java [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package no.ssb.dapla.datamaintenance.storage;
+
+import java.nio.file.Path;
+
+public record PathAndSize(Path path, Long size) {
```
Comment (takvamborgen) [inline src/main/java/no/ssb/dapla/datamaintenance/storage/PathAndSize.java:5]: Got to love `record`

File: src/test/java/no/ssb/dapla/datamaintenance/service/DataMaintenanceServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        mockDelete("/foo/bar", Map.of(
+                50, "gs://bucket50/foo/bar/50",
+                25, "gs://bucket25/foo/bar/25",
+                10, "gs://bucket10/foo/bar/10"
+        ));
+        mockFile("bucket50", "/foo/bar/50/file1", "/foo/bar/50/file2");
+        mockFile("bucket25", "/foo/bar/25/files1", "/foo/bar/25/baz/file2");
+        mockFile("bucket10", "/foo/bar/10/files1", "/foo/bar/10/baz/file2");
+
+        var delete = service.delete("/foo/bar", false)
```
Comment (takvamborgen) [inline src/test/java/no/ssb/dapla/datamaintenance/service/DataMaintenanceServiceTest.java:130]: A minor thing, but in my opinion, it is just confusing to use `var` when the type is not explicitly assigned. Here, you have to check the return type of `service.delete()` to see `delete`'s type. I do not think this improves readability.

File: src/main/java/no/ssb/dapla/datamaintenance/service/DataMaintenanceService.java (hunk 2 above)
Comment (hadrienk) [reply src/main/java/no/ssb/dapla/datamaintenance/service/DataMaintenanceService.java:164]: Good idea

**Review by takvamborgen — APPROVED** [review]

**Review by takvamborgen — APPROVED** [review]

File: pom.xml [hunk 5]
```diff
@@ -55,7 +55,17 @@
         <dependency>
             <groupId>com.google.cloud</groupId>
             <artifactId>google-cloud-nio</artifactId>
-            <version>0.122.4</version>
+            <version>0.122.6-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (hadrienk) [reply pom.xml:58]: Great! I'm changing it.

**Pull request merged**